contracts/EtomicSwap.sol

```diff
@@ -106,6 +106,36 @@ contract EtomicSwap {
         emit ReceiverSpent(_id, _secret);
     }
 
+    function watcherSpend(
+        bytes32 _id,
+        uint256 _amount,
+        bytes32 _secret,
+        address _tokenAddress,
+        address _sender,
+        address _receiver
+    ) external {
```

Comment:
As I can see, the only difference between from `receiverSpend` is that it takes an additional `_receiver` argument instead of using `msg.sender`. I'm not sure if it will be more optimal in terms of gas usage, but how about
```rust
 function receiverSpend(
        bytes32 _id,
        uint256 _amount,
        bytes32 _secret,
        address _tokenAddress,
        address _sender
    ) external {
  receiverSpendV2(_id, _amount, _secret, _tokenAddress, _sender, msg.sender)
}

 function receiverSpendV2(
        bytes32 _id,
        uint256 _amount,
        bytes32 _secret,
        address _tokenAddress,
        address _sender,
        address _receiver,
    ) external;
```
Then all new mm2 nodes will use `receiverSpendV2` including Watchers, but the old ones will use `receiverSpend`

Reply:
This looks a lot better, but it increases slightly the gas fee for the nodes using the receiverSpend method (for the takers). The increase is very little, but it can still be better to minimize the gas fee as much as possible rather than having better code, I'm also not sure. What do you think @artemii235?

Reply:
Yeah, it's more important to have lower fees in a general case. Though, I'm curious what causes gas usage increase and what's the exact difference between code variants. @caglaryucekaya Could you please provide more details on this?

Reply:
In Sergey's variant, `receiverSpend` method has an extra line with a function call to `receiverSpendV2`, which causes a slight increase in the gas usage. In our unit tests and in Remix, it increased the cost by 55 gas. Today I also though that contract size should also have an impact on the gas fee and making the contract shorter with Sergey's suggestions can actually reduce the gas fees. It does have a very slight effect, but only when the contract size exceeds certain thresholds, and in our case it didn't make any difference. So overall, together with the extra function call and reduced contract size the gas usage is again 55 more. If there's a better method to measure and compare the gas fee other than the unit tests, I can also try that.

Reply:
55 gas units is insignificant increase, so you can perform the proposed refactoring @caglaryucekaya